File: examples/snippets/modules/ROOT/examples/akkastreams-grpc-java/src/main/java/sensordata/SensorDataServiceImpl.java
package sensordata;

import cloudflow.akkastream.WritableSinkRef;
import sensordata.grpc.SensorData;
import sensordata.grpc.SensorDataService;
import sensordata.grpc.SensorReply;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public class SensorDataServiceImpl implements SensorDataService {
    private final WritableSinkRef<SensorData> sinkRef;

    public SensorDataServiceImpl(WritableSinkRef<SensorData> sinkRef) {
        this.sinkRef = sinkRef;
    }

    @Override
    public CompletionStage<SensorReply> provide(SensorData in) {
        return sinkRef.writeJava(in).thenApply(i -> SensorReply.newBuilder().setMessage("Received " + in.getPayload()).build());
    }
}

File: examples/snippets/modules/ROOT/examples/akkastreams-grpc-scala/src/main/scala/SensorDataServiceImpl.scala
package sensordata

import akka.NotUsed
import akka.stream.scaladsl.Source
import cloudflow.akkastream.WritableSinkRef

import sensordata.grpc.{ SensorData, SensorDataService, SensorReply }

import scala.concurrent.{ ExecutionContext, Future }

class SensorDataServiceImpl(sink: WritableSinkRef[SensorData])(implicit ec: ExecutionContext) extends SensorDataService {
  override def provide(in: SensorData): Future[SensorReply] = {
    println("howdy")
    sink.write(in).map(_ => { println("mapped"); SensorReply(s"Received ${in.payload}") })
  }
}
